src/lib.rs

```diff
@@ -58,60 +56,99 @@
 //! # }
 //! ```
 
-use std::iter::Peekable;
+#![no_std]
+
+use core::iter::Peekable;
 
 /// The iterator returned by `peeking_take_while`.
 ///
 /// See the [module documentation](./index.html) for details.
 pub struct PeekingTakeWhile<'a, I, P>
-    where I: 'a + Iterator
+where
+    I: Iterator,
 {
     iter: &'a mut Peekable<I>,
     predicate: P,
+    flag: bool,
 }
 
 impl<'a, I, P> Iterator for PeekingTakeWhile<'a, I, P>
-    where I: Iterator,
-          I::Item: ::std::fmt::Debug,
-          P: FnMut(&<I as Iterator>::Item) -> bool
+where
+    I: Iterator,
+    P: FnMut(&I::Item) -> bool,
 {
-    type Item = <I as Iterator>::Item;
+    type Item = I::Item;
 
+    #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        let predicate = &mut self.predicate;
-        if self.iter.peek().map_or(false, |x| !(predicate)(x)) {
+        if self.flag {
```

Comment:
I don't think we need to build fusing into this iterator. We can just defer to the nested iterator, which can be fused if necessary. Otherwise we risk redundant fusing.

Reply:
I pretty much took it from the std `TakeWhile` implementation, and I realized why it makes sense: It is necessary to guarantee that the closure is not executed again and again when checking if the iterator would return `None`, which would also happen when "extended" with further iterators. It can be a bit surprising if this is not done. It might be a good idea to simply provide two implementations: A simple one which isn't fuse, but may call the closure more often than necessary, which may lead to surprises if the closure has side-effects, and a more complex one which is fused.

Reply:
But the closure isn't called unless the underlying iterator returns `Some`, so I don't see how this would be more often than necessary.

Reply:
yes, but because it is a `take_while` implementation, the underlying iterator could return `Some` (causing the `peek()` to pick that up), but the `next_if` invocation doesn't forward it, because the predicate returns `false`. On the next call to `next()` of the `take_while` iterator, the predicate would be called again, because the underlying `Peekable` instance still holds the same element, as an unconditional `next()` on the `Peekable` instance was never called.